File: Render Engine/RenderPhases.cs
using System.Collections.Generic;
using System.IO;
using System;

namespace SETUE
{
    public class RenderPhase
    {
        public int PhaseId;
        public string CommandName = "";
        public string? PipelineId;
        public bool Enabled;
        public string DependencyId = "None";
        public string Description = "";
    }

    public static class RenderPhases
    {
        public static readonly List<RenderPhase> Phases = new();

        public static void Load(string path = "Shaders/RenderPhases.csv")
        {
            Console.WriteLine($"[RenderPhases] Loading from '{path}' (full path: {Path.GetFullPath(path)})");
            Phases.Clear();
            
            if (!File.Exists(path))
            {
                Console.WriteLine($"[RenderPhases] ERROR: File not found: {path}");
                return;
            }

            var lines = File.ReadAllLines(path);
            Console.WriteLine($"[RenderPhases] Read {lines.Length} lines");
            
            for (int i = 1; i < lines.Length; i++)
            {
                if (string.IsNullOrWhiteSpace(lines[i]))
                    continue;

                var row = lines[i].Split(',');
                if (row.Length < 6)
                {
                    Console.WriteLine($"[RenderPhases] Skipping row {i}: only {row.Length} columns");
                    continue;
                }

                var phase = new RenderPhase
                {
                    PhaseId = int.Parse(row[0]),
                    CommandName = row[1],
                    PipelineId = string.IsNullOrWhiteSpace(row[2]) ? null : row[2],
                    Enabled = bool.Parse(row[3]),
                    DependencyId = row[4],
                    Description = row[5]
                };
                Phases.Add(phase);
                Console.WriteLine($"[RenderPhases] Loaded phase: {phase.CommandName} Enabled={phase.Enabled}");
            }
            Console.WriteLine($"[RenderPhases] Total phases loaded: {Phases.Count}");
        }
    }
}

File: Render Engine/RenderPhases.ts
// RenderPhases.ts
import { parseCsvArray } from './parseCsv.js';
import { registerSystemMethod } from './Core/Scheduler.js';
import * as PIXI from 'pixi.js';

let stage: PIXI.Container | null = null;
let uiContainer: PIXI.Container | null = null;
let textContainer: PIXI.Container | null = null;

export function setRenderContainers(
  mainStage: PIXI.Container,
  uiLayer: PIXI.Container,
  textLayer: PIXI.Container
) {
  stage = mainStage;
  uiContainer = uiLayer;
  textContainer = textLayer;
}

interface PhaseDef {
  PhaseId: number;
  CommandName: string;
  DependencyId: number;
  Enabled: boolean;
}

let phases: PhaseDef[] = [];

async function Load() {
  const response = await fetch('RenderPhases.csv');
  if (!response.ok) {
    console.warn('[RenderPhases] RenderPhases.csv not found');
    return;
  }
  const csvText = await response.text();
  const rows = parseCsvArray(csvText);

  phases = rows
    .filter(r => r['Enabled']?.toLowerCase() === 'true')
    .map(r => ({
      PhaseId: parseInt(r['PhaseId'] || '0'),
      CommandName: r['CommandName'] || '',
      DependencyId: r['DependencyId'] && r['DependencyId'] !== 'None' ? parseInt(r['DependencyId']) : 0,
      Enabled: true,
    }))
    .sort((a, b) => a.PhaseId - b.PhaseId);

  console.log(`[RenderPhases] Loaded ${phases.length} phases`);
}

function Update() {
  for (const phase of phases) {
    switch (phase.CommandName) {
      case 'Begin_Frame':
        // PixiJS ticker handles this
        break;
      case 'Draw_Meshes':
        // Placeholder for 3D meshes
        break;
      case 'Draw_UI':
        if (uiContainer && stage) {
          stage.setChildIndex(uiContainer, stage.children.length - 1);
        }
        break;
      case 'Draw_Text':
        if (textContainer && stage) {
          stage.setChildIndex(textContainer, stage.children.length - 1);
        }
        break;
      case 'End_Frame':
        // Finalisation
        break;
      default:
        break;
    }
  }
}

registerSystemMethod('SETUE.RenderPhases', 'Load', Load);
registerSystemMethod('SETUE.RenderPhases', 'Update', Update);
